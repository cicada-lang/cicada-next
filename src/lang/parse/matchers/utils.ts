import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
