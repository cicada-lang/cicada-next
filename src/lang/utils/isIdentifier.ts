const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
