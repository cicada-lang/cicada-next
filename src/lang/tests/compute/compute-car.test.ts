import { expect, test } from "vitest"
import { runCode } from "../utils"

test("compute Car", async () => {
  const output = await runCode(`

let pair: exists (Type) Type = cons(Type, Type)
compute car(pair)

`)

  expect(output).toMatchSnapshot()
})

test("compute Car -- just car", async () => {
  const output = await runCode(`

compute car

`)

  expect(output).toMatchSnapshot()
})

test("compute Car -- my_car", async () => {
  const output = await runCode(`

function my_car(
  implicit A: Type,
  implicit B: (x: A) -> Type,
  pair: exists (x: A) B(x),
): A {
  return car(pair)
}

compute my_car(cons("a", "b"))

`)

  expect(output).toMatchSnapshot()
})
