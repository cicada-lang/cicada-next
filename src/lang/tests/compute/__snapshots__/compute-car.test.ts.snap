// Vitest Snapshot v1, https://bun:test.dev/guide/snapshot.html

exports[`compute Car -- just car 1`] = `"(implicit A, implicit B, target) => car(target): (implicit A: Type, implicit B: (A) -> Type, exists (A) B(x)) -> A"`;

exports[`compute Car -- my_car 1`] = `""a": String"`;

exports[`compute Car 1`] = `"Type: Type"`;
